homeassistant/components/dlna_dms/dms.py

```diff
@@ -0,0 +1,715 @@
+from typing import Any, TypeVar, cast
+
+from async_upnp_client import UpnpEventHandler, UpnpFactory, UpnpRequester
+from async_upnp_client.aiohttp import AiohttpSessionRequester
+from async_upnp_client.const import NotificationSubType
+from async_upnp_client.exceptions import UpnpActionError, UpnpConnectionError, UpnpError
+from async_upnp_client.profiles.dlna import ContentDirectoryErrorCode, DmsDevice
+from didl_lite import didl_lite
+
+from homeassistant.components import ssdp
+from homeassistant.components.media_player.const import MEDIA_CLASS_DIRECTORY
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.components.media_source.error import Unresolvable
+from homeassistant.components.media_source.models import BrowseMediaSource, PlayMedia
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_DEVICE_ID, CONF_ENTITY_ID, CONF_URL
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import aiohttp_client
+
+from .const import (
+    DLNA_BROWSE_FILTER,
+    DLNA_PATH_FILTER,
+    DLNA_RESOLVE_FILTER,
+    DLNA_SORT_CRITERIA,
+    DOMAIN,
+    LOGGER,
+    MEDIA_CLASS_MAP,
+    MEDIA_TYPE_MAP,
+    PATH_OBJECT_ID_FLAG,
+    PATH_SEARCH_FLAG,
+    PATH_SEP,
+    ROOT_OBJECT_ID,
+    STREAMABLE_PROTOCOLS,
+)
+
+_DmsEntityMethod = TypeVar("_DmsEntityMethod", bound="DmsEntity")
+_RetType = TypeVar("_RetType")
+
+
+class DlnaDmsData:
+    """Storage class for domain global data."""
+
+    hass: HomeAssistant
+    lock: asyncio.Lock
+    requester: UpnpRequester
+    upnp_factory: UpnpFactory
+    event_handler: UpnpEventHandler
+    entities: dict[str, DmsEntity]  # Indexed by unique_id
+    sources: dict[str, DmsEntity]  # Indexed by source_id
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+    ) -> None:
+        """Initialize global data."""
+        self.hass = hass
+        self.lock = asyncio.Lock()
+        session = aiohttp_client.async_get_clientsession(hass, verify_ssl=False)
+        self.requester = AiohttpSessionRequester(session, with_sleep=True)
+        self.upnp_factory = UpnpFactory(self.requester, non_strict=True)
+        # NOTE: event_handler is not actually used, and is only created to
+        # satisfy the DmsDevice.__init__ signature
+        self.event_handler = UpnpEventHandler("", self.requester)
+        self.entities = {}
+        self.sources = {}
+
+    async def async_setup_entry(self, config_entry: ConfigEntry) -> bool:
+        """Create a DMS device connection from a config entry."""
+        entity = DmsEntity(self.hass, config_entry)
+        async with self.lock:
+            if entity.source_id in self.sources:
+                LOGGER.error("The source_id %s is already in use", entity.source_id)
+                return False
+            self.entities[entity.unique_id] = entity
+            self.sources[entity.source_id] = entity
+        await entity.async_added_to_hass()
```

Comment:
Ah, this is not a Home Assistant entity object.

Can you rename the class. I was confused because the called functions have same names as normal entities.

Reply:
I can see how that's confusing. I've renamed it now.

For background, this was originally based on the idea of [`MediaSourceEntity`](https://github.com/home-assistant/architecture/discussions/675), which was rejected because media sources haven't got states to report or services to call, unlike other HA entities.